File: packages/core/src/receipt.js
import { randomBytes } from 'node:crypto';
import { canonicalBytes } from './canonical.js';
import { sha256, hashObject, hex, unhex, RECEIPT_PREFIX, equalBytes } from './hash.js';
import { leafHash } from './merkle.js';
import { sign, verify } from './keys.js';
import { redact } from './redact.js';

/**
 * The receipt: one signed, chained record of one thing an agent did.
 *
 * Three properties are load-bearing, and each exists because a specific
 * dispute is foreseeable:
 *
 *   signed    "your logs say X" → the signature says a specific key asserted
 *             X, and the private key never leaves the agent runtime.
 *   chained   every receipt commits to its predecessor's hash, so removing or
 *             reordering a single entry invalidates everything after it.
 *   sealed    arguments and results are stored as salted commitments plus a
 *             redacted preview. The log can be handed to an outside auditor
 *             without handing over customer data — and the holder of the
 *             original can still prove, later, exactly what it was.
 */

/** Wire format version. Bump only for breaking changes. */
export const RECEIPT_VERSION = 1;

/** `prev` for the first entry in a log. */
export const GENESIS_PREV = '0'.repeat(64);

/**
 * @typedef {object} Sealed
 * @property {string} hash  sha256(salt || canonical(value)), hex.
 * @property {number} size  Byte length of the canonical form — cheap signal of scale.
 * @property {unknown} [preview]  Redacted copy, safe to publish.
 * @property {string[]} [redacted]  Types that were masked, e.g. `["email"]`.
 */

/**
 * Commit to a value without storing it.
 *
 * The salt is what makes this worth doing. A bare hash of a low-entropy value
 * — `{"amount": 50}`, a customer's email — is trivially brute-forced, so an
 * unsalted commitment leaks exactly the data it pretends to protect.
 *
 * The salt is returned *alongside* the commitment rather than inside it, and
 * it is never part of what gets signed. Two consequences, both deliberate:
 *
 *   - A receipt is safe to publish as-is. There is no "remember to strip the
 *     secrets before exporting" step to forget.
 *   - Destroying the salts renders the commitments permanently un-openable
 *     while leaving every signature and chain link intact. That is a real
 *     erasure — a GDPR Article 17 request can be honoured without gutting the
 *     audit trail, which is otherwise a direct conflict between two duties.
 *
 * @param {unknown} value
 * @param {object} [opts]
 * @param {boolean} [opts.preview=true]  Include a redacted copy.
 * @returns {{ sealed: Sealed, salt: string }}
 */
export function seal(value, opts = {}) {
  const salt = randomBytes(16);
  const bytes = canonicalBytes(value ?? null);
  /** @type {Sealed} */
  const sealed = {
    hash: hex(sha256(salt, bytes)),
    size: bytes.length,
  };
  if (opts.preview !== false) {
    const { value: masked, findings } = redact(value ?? null);
    sealed.preview = masked;
    if (findings.length > 0) {
      sealed.redacted = [...new Set(findings.map((f) => f.type))].sort();
    }
  }
  return { sealed, salt: salt.toString('base64url') };
}

/**
 * Prove that `value` is what a sealed commitment referred to.
 *
 * @param {Sealed} sealed
 * @param {string|undefined} saltB64u  From the log's salt store. Absent means
 *   the payload was crypto-shredded and can no longer be opened by anyone.
 * @param {unknown} value
 * @returns {boolean}
 */
export function openSeal(sealed, saltB64u, value) {
  if (!sealed || typeof saltB64u !== 'string') return false;
  let salt;
  try {
    salt = Buffer.from(saltB64u, 'base64url');
  } catch {
    return false;
  }
  const expect = sha256(salt, canonicalBytes(value ?? null));
  try {
    return equalBytes(expect, unhex(sealed.hash));
  } catch {
    return false;
  }
}

/**
 * @typedef {object} Actor
 * @property {string} agent      Model or agent identifier, e.g. `claude-opus-5`.
 * @property {string} runtime    What produced the receipt, e.g. `proofwire-proxy/0.1.0`.
 * @property {string} session    Groups the receipts of one agent run.
 * @property {string} principal  Whose authority the agent acted under.
 */

/**
 * @typedef {object} Decision
 * @property {'allow'|'deny'|'escalate'} outcome
 * @property {string} policy     Hash of the policy document in force.
 * @property {string[]} rules    Rule ids that fired, in evaluation order.
 * @property {string} [reason]   Human-readable justification.
 * @property {{ by: string, at: string, note?: string }} [approval]
 */

/**
 * @typedef {object} ReceiptBody
 * @property {number} v
 * @property {string} log
 * @property {number} seq
 * @property {string} prev
 * @property {string} ts
 * @property {'atomic'|'intent'|'outcome'} phase
 * @property {string} [ref]
 * @property {Actor} actor
 * @property {{ kind: string, target: string, metrics?: Record<string, number>, params: Sealed }} action
 * @property {Decision} decision
 * @property {null | { status: 'ok'|'error', code?: string, latencyMs?: number, payload: Sealed }} result
 */

/**
 * @typedef {ReceiptBody & { attest: { alg: 'ed25519', kid: string, sig: string } }} Receipt
 */

/**
 * Assemble an unsigned receipt body.
 *
 * @param {object} args
 * @param {string} args.log
 * @param {number} args.seq
 * @param {string} args.prev
 * @param {Actor} args.actor
 * @param {{ kind: string, target: string, params: unknown, metrics?: Record<string, number> }} args.action
 * @param {Decision} args.decision
 * @param {null | { status: 'ok'|'error', code?: string, latencyMs?: number, payload?: unknown }} [args.result]
 * @param {'atomic'|'intent'|'outcome'} [args.phase]
 * @param {string} [args.ref]
 * @param {string} [args.ts]
 * @returns {{ body: ReceiptBody, salts: { params: string, result?: string } }}
 *   The salts are the caller's to store separately — see `seal`.
 */
export function buildReceipt(args) {
  const params = seal(args.action.params);
  const payload = args.result ? seal(args.result.payload) : null;

  /** @type {ReceiptBody} */
  const body = {
    v: RECEIPT_VERSION,
    log: args.log,
    seq: args.seq,
    prev: args.prev,
    ts: args.ts ?? new Date().toISOString(),
    phase: args.phase ?? 'atomic',
    actor: args.actor,
    action: {
      kind: args.action.kind,
      target: args.action.target,
      // Metrics stay in the clear by design: budgets are aggregated from them,
      // and a spend cap you cannot recompute from the log is not auditable.
      // Only non-sensitive numbers belong here.
      ...(args.action.metrics ? { metrics: args.action.metrics } : {}),
      params: params.sealed,
    },
    decision: args.decision,
    result:
      args.result && payload
        ? {
            status: args.result.status,
            ...(args.result.code !== undefined ? { code: args.result.code } : {}),
            ...(args.result.latencyMs !== undefined ? { latencyMs: args.result.latencyMs } : {}),
            payload: payload.sealed,
          }
        : null,
  };
  if (args.ref) body.ref = args.ref;

  return {
    body,
    salts: { params: params.salt, ...(payload ? { result: payload.salt } : {}) },
  };
}

/**
 * The bytes a signature covers: the body, canonicalized, under the receipt tag.
 *
 * @param {ReceiptBody} body
 * @returns {Buffer}
 */
export function receiptDigest(body) {
  const { attest, ...rest } = /** @type {Record<string, unknown>} */ (body);
  return hashObject(RECEIPT_PREFIX, rest);
}

/**
 * The receipt's identity — and, deliberately, its Merkle leaf hash.
 *
 * Reusing one value for the chain pointer and the tree leaf means an inclusion
 * proof and a chain link cannot disagree about which receipt they refer to.
 *
 * @param {Receipt} receipt
 * @returns {string} hex
 */
export function entryHash(receipt) {
  return hex(leafHash(canonicalBytes(receipt)));
}

/**
 * @param {import('./keys.js').Identity} identity
 * @param {ReceiptBody} body
 * @returns {Receipt}
 */
export function signReceipt(identity, body) {
  const sig = sign(identity, receiptDigest(body));
  return { ...body, attest: { alg: 'ed25519', kid: identity.kid, sig } };
}

/**
 * @typedef {Record<string, string>} Keyring  kid → base64url public key.
 */

/**
 * @typedef {object} VerifyIssue
 * @property {number} [seq]
 * @property {'signature'|'chain'|'sequence'|'format'|'key'|'time'} kind
 * @property {string} message
 */

/**
 * Check one receipt's signature against a keyring.
 *
 * @param {Receipt} receipt
 * @param {Keyring} keyring
 * @returns {VerifyIssue[]}  Empty when the receipt is sound.
 */
export function verifyReceipt(receipt, keyring) {
  /** @type {VerifyIssue[]} */
  const issues = [];
  const seq = receipt?.seq;

  if (!receipt || typeof receipt !== 'object') {
    return [{ kind: 'format', message: 'receipt is not an object' }];
  }
  if (receipt.v !== RECEIPT_VERSION) {
    issues.push({ seq, kind: 'format', message: `unsupported receipt version ${receipt.v}` });
  }
  const attest = receipt.attest;
  if (!attest || attest.alg !== 'ed25519' || typeof attest.sig !== 'string') {
    issues.push({ seq, kind: 'format', message: 'missing or malformed attestation' });
    return issues;
  }
  const pub = keyring[attest.kid];
  if (!pub) {
    issues.push({ seq, kind: 'key', message: `no public key for kid ${attest.kid}` });
    return issues;
  }
  const { attest: _drop, ...body } = receipt;
  if (!verify(pub, receiptDigest(/** @type {ReceiptBody} */ (body)), attest.sig)) {
    issues.push({ seq, kind: 'signature', message: `signature does not verify for kid ${attest.kid}` });
  }
  return issues;
}

/**
 * Verify a contiguous run of receipts: signatures, sequence numbers, chain
 * links, and monotonic timestamps.
 *
 * Timestamp regressions are reported but treated as a weaker signal than the
 * rest — clocks genuinely do step backwards, whereas a broken chain link
 * cannot happen by accident.
 *
 * @param {Receipt[]} receipts  Oldest first.
 * @param {Keyring} keyring
 * @param {object} [opts]
 * @param {string} [opts.expectPrev]  Chain hash the run should start from.
 * @param {number} [opts.expectSeq]   Sequence the run should start at.
 * @returns {{ ok: boolean, issues: VerifyIssue[], head: string, count: number }}
 */
export function verifyChain(receipts, keyring, opts = {}) {
  /** @type {VerifyIssue[]} */
  const issues = [];
  let prev = opts.expectPrev ?? GENESIS_PREV;
  let expectedSeq = opts.expectSeq ?? 0;
  let lastTs = null;

  for (const receipt of receipts) {
    issues.push(...verifyReceipt(receipt, keyring));

    if (receipt.seq !== expectedSeq) {
      issues.push({
        seq: receipt.seq,
        kind: 'sequence',
        message: `expected seq ${expectedSeq}, found ${receipt.seq} — entries missing or reordered`,
      });
    }
    if (receipt.prev !== prev) {
      issues.push({
        seq: receipt.seq,
        kind: 'chain',
        message: `chain break: prev is ${String(receipt.prev).slice(0, 12)}…, expected ${prev.slice(0, 12)}…`,
      });
    }
    if (lastTs && receipt.ts < lastTs) {
      issues.push({
        seq: receipt.seq,
        kind: 'time',
        message: `timestamp ${receipt.ts} precedes the previous entry's ${lastTs}`,
      });
    }

    lastTs = receipt.ts;
    prev = entryHash(receipt);
    expectedSeq = receipt.seq + 1;
  }

  return { ok: issues.length === 0, issues, head: prev, count: receipts.length };
}

File: sdk/python/src/proof_wire/receipt.py
"""Receipts: build, seal, sign, verify. Mirrors packages/core/src/receipt.js.

A receipt's signature covers ``SHA-256(0x02 || canonical(body without attest))``,
and its place in the log's Merkle tree is ``SHA-256(0x00 || canonical(receipt))``.
Arguments and results are *sealed*: a salted hash commits to them, and only a
redacted preview is stored in the clear.
"""

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any, Optional

from .canonical import canonical_bytes
from .hashing import RECEIPT_PREFIX, b64url, equal_bytes, from_b64url, from_hex, hash_object, sha256, to_hex
from .keys import Identity, sign, verify
from .merkle import leaf_hash
from .redact import redact

RECEIPT_VERSION = 1
GENESIS_PREV = "0" * 64
PHASES = ("atomic", "intent", "outcome")


def now_iso() -> str:
    """UTC, millisecond precision, ``Z``: the same shape as JavaScript's toISOString()."""
    t = datetime.now(timezone.utc)
    return t.strftime("%Y-%m-%dT%H:%M:%S.") + f"{t.microsecond // 1000:03d}Z"


def seal(value: Any, preview: bool = True) -> tuple[dict, str]:
    """Commit to a value without storing it.

    Returns the sealed form that goes into the receipt, and the salt, which
    goes to the log's separate salts file. Destroying the salt later makes
    the value permanently unprovable while the receipt still verifies.
    """
    salt = os.urandom(16)
    data = canonical_bytes(value)
    sealed: dict[str, Any] = {"hash": to_hex(sha256(salt, data)), "size": len(data)}
    if preview:
        masked, findings = redact(value)
        sealed["preview"] = masked
        if findings:
            sealed["redacted"] = sorted({f.type for f in findings})
    return sealed, b64url(salt)


def open_seal(sealed: dict, salt: Optional[str], value: Any) -> bool:
    """Whether ``value`` is what the seal committed to."""
    if not sealed or not isinstance(salt, str):
        return False
    try:
        expect = sha256(from_b64url(salt), canonical_bytes(value))
        return equal_bytes(expect, from_hex(sealed["hash"]))
    except (ValueError, KeyError, TypeError):
        return False


def _require(value: Any, field: str) -> None:
    if not isinstance(value, str) or not value:
        raise ValueError(f"build_receipt: {field} is required and must be a non-empty string")


def build_receipt(
    *,
    log: str,
    seq: int,
    prev: str,
    actor: dict,
    action: dict,
    decision: dict,
    result: Optional[dict] = None,
    phase: str = "atomic",
    ref: Optional[str] = None,
    ts: Optional[str] = None,
) -> tuple[dict, dict]:
    """An unsigned receipt body, and the salts to keep beside it.

    ``action`` is ``{kind, target, params, metrics?}``; ``result`` is
    ``{status, code?, latencyMs?, payload}``.
    """
    for path, v in (
        ("actor.principal", actor.get("principal")),
        ("actor.agent", actor.get("agent")),
        ("actor.session", actor.get("session")),
        ("action.kind", action.get("kind")),
        ("action.target", action.get("target")),
        ("decision.outcome", decision.get("outcome")),
    ):
        _require(v, path)
    if phase not in PHASES:
        raise ValueError(f"phase must be one of {', '.join(PHASES)}")

    params_sealed, params_salt = seal(action.get("params"))
    act: dict[str, Any] = {"kind": action["kind"], "target": action["target"]}
    # Metrics stay in the clear: budgets are computed from them, and a cap
    # nobody can recompute from the log is not auditable.
    if action.get("metrics"):
        act["metrics"] = action["metrics"]
    act["params"] = params_sealed

    salts: dict[str, str] = {"params": params_salt}
    res: Optional[dict] = None
    if result is not None:
        payload_sealed, payload_salt = seal(result.get("payload"))
        salts["result"] = payload_salt
        res = {"status": result["status"]}
        if result.get("code") is not None:
            res["code"] = result["code"]
        if result.get("latencyMs") is not None:
            res["latencyMs"] = result["latencyMs"]
        res["payload"] = payload_sealed

    body: dict[str, Any] = {
        "v": RECEIPT_VERSION,
        "log": log,
        "seq": seq,
        "prev": prev,
        "ts": ts or now_iso(),
        "phase": phase,
        "actor": actor,
        "action": act,
        "decision": decision,
        "result": res,
    }
    if ref:
        body["ref"] = ref
    return body, salts


def receipt_digest(body: dict) -> bytes:
    return hash_object(RECEIPT_PREFIX, {k: v for k, v in body.items() if k != "attest"})


def entry_hash(receipt: dict) -> str:
    """The receipt's leaf in the log's Merkle tree, and the next receipt's ``prev``."""
    return to_hex(leaf_hash(canonical_bytes(receipt)))


def sign_receipt(identity: Identity, body: dict) -> dict:
    return {**body, "attest": {"alg": "ed25519", "kid": identity.kid, "sig": sign(identity, receipt_digest(body))}}


def verify_receipt(receipt: Any, keyring: dict) -> list[dict]:
    """Problems with one receipt, as ``{seq, kind, message}``; empty means valid."""
    if not isinstance(receipt, dict):
        return [{"kind": "format", "message": "receipt is not an object"}]
    seq = receipt.get("seq")
    issues: list[dict] = []
    if receipt.get("v") != RECEIPT_VERSION:
        issues.append({"seq": seq, "kind": "format", "message": f"unsupported receipt version {receipt.get('v')}"})
    actor = receipt.get("actor") or {}
    action = receipt.get("action") or {}
    decision = receipt.get("decision") or {}
    for path, v in (
        ("ts", receipt.get("ts")),
        ("actor.principal", actor.get("principal")),
        ("actor.agent", actor.get("agent")),
        ("actor.session", actor.get("session")),
        ("action.kind", action.get("kind")),
        ("action.target", action.get("target")),
        ("decision.outcome", decision.get("outcome")),
    ):
        if not isinstance(v, str) or not v:
            issues.append({"seq": seq, "kind": "format", "message": f"{path} is required and must be a non-empty string"})
    if receipt.get("phase") not in PHASES:
        issues.append({"seq": seq, "kind": "format", "message": f"phase must be one of atomic, intent, outcome; got {receipt.get('phase')!r}"})
    attest = receipt.get("attest")
    if not isinstance(attest, dict) or attest.get("alg") != "ed25519" or not isinstance(attest.get("sig"), str):
        issues.append({"seq": seq, "kind": "format", "message": "missing or malformed attestation"})
        return issues
    pub = keyring.get(attest.get("kid"))
    if not pub:
        issues.append({"seq": seq, "kind": "key", "message": f"no public key for kid {attest.get('kid')}"})
        return issues
    if not verify(pub, receipt_digest(receipt), attest["sig"]):
        issues.append({"seq": seq, "kind": "signature", "message": f"signature does not verify for kid {attest.get('kid')}"})
    return issues


def verify_chain(receipts: list[dict], keyring: dict, expect_prev: str = GENESIS_PREV, expect_seq: int = 0) -> dict:
    """Signatures, sequence, hash chain and timestamp order, end to end."""
    issues: list[dict] = []
    prev, expected, last_ts = expect_prev, expect_seq, None
    for r in receipts:
        issues.extend(verify_receipt(r, keyring))
        seq = r.get("seq")
        if seq != expected:
            issues.append({"seq": seq, "kind": "sequence", "message": f"expected seq {expected}, found {seq}: entries missing or reordered"})
        if r.get("prev") != prev:
            issues.append({"seq": seq, "kind": "chain", "message": f"chain break: prev is {str(r.get('prev'))[:12]}…, expected {prev[:12]}…"})
        ts = r.get("ts")
        if last_ts and isinstance(ts, str) and ts < last_ts:
            issues.append({"seq": seq, "kind": "time", "message": f"timestamp {ts} precedes the previous entry's {last_ts}"})
        last_ts = ts
        prev = entry_hash(r)
        expected = (seq if isinstance(seq, int) else expected) + 1
    return {"ok": not issues, "issues": issues, "head": prev, "count": len(receipts)}
